db/src/db.rs

```diff
@@ -416,45 +437,438 @@ impl DB {
+
+        Ok(())
+    }
+
+    /// Update the contents of the `datoms` materialized view with the new transaction.
+    ///
+    /// This applies the contents of `search_results` to the `datoms` table (in place).
+    ///
+    /// See https://github.com/mozilla/mentat/wiki/Transacting:-entity-to-SQL-translation.
+    // TODO: capture `conn` in a `TxInternal` structure.
+    fn update_datoms(&self, conn: &rusqlite::Connection, tx: Entid) -> Result<()> {
+        // Delete datoms that were retracted, or those that were :db.cardinality/one and will be
+        // replaced.
+        let s = r#"
+            WITH ids AS (SELECT rid
+                         FROM temp.search_results
+                         WHERE rid IS NOT NULL AND
+                               ((added0 IS 0) OR
+                                (added0 IS 1 AND search_type IS ':db.cardinality/one' AND v0 IS NOT v)))
+            DELETE FROM datoms WHERE rowid IN ids"#;
+
+        let mut stmt = conn.prepare_cached(s)?;
+        stmt.execute(&[])
+            .map(|_c| ())
+            .chain_err(|| "Could not update datoms: failed to retract datoms already present")?;
+
+        // Insert datoms that were added and not already present.
+        let s = r#"
+          INSERT INTO datoms (e, a, v, tx, value_type_tag, index_avet, index_vaet, index_fulltext, unique_value)
+          SELECT e0, a0, v0, ?, value_type_tag0,
+                 index_avet0, index_vaet0, index_fulltext0, unique_value0
+          FROM temp.search_results
+          WHERE added0 IS 1 AND ((rid IS NULL) OR ((rid IS NOT NULL) AND (v0 IS NOT v)))"#;
+
+        let mut stmt = conn.prepare_cached(s)?;
+        stmt.execute(&[&tx])
+            .map(|_c| ())
+            .chain_err(|| "Could not update datoms: failed to add datoms not already present")?;
+
+        Ok(())
+    }
+
+    /// Transact the given `entities` against the given SQLite `conn`, using the metadata in
+    /// `self.DB`.
+    ///
+    /// This approach is explained in https://github.com/mozilla/mentat/wiki/Transacting.
     // TODO: move this to the transactor layer.
-    pub fn transact_internal(&self, conn: &rusqlite::Connection, entities: &[Entity]) -> Result<()>{
-        // TODO: manage :db/tx, write :db/txInstant.
-        let tx = 1;
+    pub fn transact_internal(&self, conn: &rusqlite::Connection, entities: &[Entity], tx: Entid) -> Result<()>{
+        // TODO: push these into an internal transaction report?
+
+        /// Assertions that are :db.cardinality/one and not :db.fulltext.
+        let mut non_fts_one: Vec<ReducedEntity> = vec![];
+
+        /// Assertions that are :db.cardinality/many and not :db.fulltext.
+        let mut non_fts_many: Vec<ReducedEntity> = vec![];
+
+        // Transact [:db/add :db/txInstant NOW :db/tx].
+        // TODO: allow this to be present in the transaction data.
+        let now = time::get_time();
+        let tx_instant = (now.sec as i64 * 1_000) + (now.nsec as i64 / (1_000_000));
```

Comment:
Please extract a `now_millis()` utility function. We'll need this elsewhere, I imagine, and it'd be nice to test it.